Recycler/function composition

ghostbuster91 opened the pull request:
This approach let us avoid unnecessary class creation. I think it's more elegant way, but let me know what do you think.

```

fun createPersonItemAdapter(person: Person) =
        StableItemAdapter(123, ::PersonViewHolder, person, person.hashCode().toLong()) { v ->
            v.nameView.text = name
            v.lastNameView.text = name
        }

data class Person(val name: String, val lastName: String)

class PersonViewHolder(view: View) : RecyclerView.ViewHolder(view) {
    val nameView = TextView(view.context)
    val lastNameView = TextView(view.context)
}
```

ghostbuster91 committed:
Add StableRecyclerViewAdapter, this closes #4

ghostbuster91 committed:
Replace inheritance with function composition

OrdonTeam commented:
I don't get why ItemAdapterImpl needs generic type T.
Please also add some dummy tests as an usage documentation.

OrdonTeam reviewed (changes requested):
Try to remove generic type T from ItemAdapterImpl

ghostbuster91 commented:
I did it on purpose, to make `bind` a pure function without variables capture.
Without this,`binder` had to be declared together with `StableItemAdapter` because both of them need access to data object ( `StableItemAdapter` needs it because it takes `stableId` parameter which probably will be calculated based on data object).

Another advantage is that `bind` function can be declared now as function on `T` class, so we have build-in with function.

mg6maciej commented:
I don't like the proposed changes for a few reasons.
Most important reason is that any code that is added to commons should be already battle-tested in at least 3 projects. The kind of code that we copied to new project from old one without (too many) modifications. I just assumed here, that @ghostbuster91 started using stable ids in the current or previous project.
Another thing is hiding all the details from client code. With a simple `PersonItemAdapter(person)` you don't need to care about all the details of `ViewHolder` implementation or usage, i.e. binding, getting id from person or having the proper `R.layout.person_item` in `View` implementation code. Proposed approach also has problems with reusability. Remember `DoctorItemAdapter` and friends?
Less important, but I think you got some wrong assumptions. `T.(VH) -> Unit` is not a pure function. And you would never want to define such function directly on T (e.g. `Person`), because it depends on `ViewHolder`, which depend on `View`, which is part of Android framework. If anything, I'd much rather see it defined as `VH.(T) -> Unit`, but still that's less readable than defining a method in a separate class.

ghostbuster91 commented:
Thanks for replay.
Let me address those issues one by one.

> Most important reason is that any code that is added to commons should be already battle-tested in at least 3 projects. The kind of code that we copied to new project from old one without (too many) modifications. 

 Partially agree. The code in this library for sure should be battle tested and I admit that I didn't have a chance to test my changes in at least 3 projects. But I implemented this approach (functions over classes) in similar circumstances and it worked well, so this is why I wanted to share this code with you.

> I just assumed here, that @ghostbuster91 started using stable ids in the current or previous project.

StableIds is another topic so leave it away.

> Another thing is hiding all the details from client code. With a simple PersonItemAdapter(person) you don't need to care about all the details of ViewHolder implementation or usage, i.e. binding, getting id from person or having the proper R.layout.person_item in View implementation code. 

I don't agree. You can have this function in separate file, and from client code you will have similar perspective. You just only replace `PersonItemAdapter(person)` with function call `createPersonItemAdapter(person)`, or am I missing something?

> Proposed approach also has problems with reusability. Remember DoctorItemAdapter and friends

Don't agree on this either. Because `bind` is a function is way simpler to compose many functions and to reuse them from different contexts. Probably you get it wrong because what I said about `bind` function earlier, I will try to explain it in next paragraph.

> Less important, but I think you got some wrong assumptions. T.(VH) -> Unit is not a pure function. And you would never want to define such function directly on T (e.g. Person), because it depends on ViewHolder, which depend on View, which is part of Android framework. If anything, I'd much rather see it defined as VH.(T) -> Unit, but still that's less readable than defining a method in a separate class.

To explain this I will refer to what @OrdonTeam said. 
I made itemAdapter accept additional parameter and have two generics to achieve following change:
Without additional parameter:

```
fun createPersonItemAdapter(person: Person): StableItemAdapter<Person, PersonViewHolder> {
    val bind = { v: PersonViewHolder ->
        v.nameView.text = person.name
        v.lastNameView.text = person.lastName
    }
    return StableItemAdapter(123, ::PersonViewHolder, person.hashCode().toLong(), bind)
}
```

The person variable is from outer function so it's captured by bind function and in such approach re-usability might be impossible. So I changed it to accept two parameters, both person and personViewHolder, which resulted in:

```
fun createPersonItemAdapter(person: Person) : StableItemAdapter<Person, PersonViewHolder>{
    val binder = { v:PersonViewHolder,p: Person ->
        v.nameView.text = p.name
        v.lastNameView.text = p.lastName
    }
    return StableItemAdapter(123, ::PersonViewHolder, person, person.hashCode().toLong(), binder)
}
```

But I also saw that within bind function (also in 'old' approach) we use `with` function to avoid referring to object by name so many times. In new approach this could look like this:

```
val binder = { v:PersonViewHolder,p: Person ->
       with(p){
           v.nameView.text = p.name
           v.lastNameView.text = p.lastName
      }
    }
```

As soon as I saw it I started thinking that it is also a boiler plate so I changed the signature of bind function to that which you already saw. So I far from forcing you to implement function in your data class which will have a dependency to android framework (Of course we can invert it so it will be a function on viewHolder which will accept data class)

I hope that now everything is clear, if not feel free to ask.

ghostbuster91 commented:
This approach has performance issues because it creates many anonymous classes.
